torch/csrc/jit/script/convert_to_ssa.cpp

```diff
@@ -118,20 +118,25 @@ struct ControlFlowLoadStores {
     auto loop_vars = addControlFlowLoadStores(body_block);
 
     for (const auto& name : loop_vars->definedVariables()) {
-      // we require that the variable is defined outside the loop to be emitted,
-      // and we do not refine the type of the parent variable since the loop may
-      // not be entered.
+      // if the variable local to the loop body, then
+      // we do not need a loop carried variabl for it
       auto parent_type = environment_stack->findInAnyFrame(name);
       if (!parent_type) {
         continue;
       }
 
+      // since the loop may execute 0 or many times, the output types
+      // of the loop and the input loop carried dependencies are conservatively
+      // the union of the output of the body and the input to the loop
+      auto block_type = loop_vars->findInThisFrame(name);
+      auto unified_type = unifyTypes(parent_type, block_type).value();
```

Comment:
Why do we have to unify `parent_type` and `block_type` here ? In compiler.cpp there isn't a similar logic; we enforce that the loop block variables subtype the parent type and when the loop is done emitting the variable remains the parent type.

I think  having a difference in type between the type environment in compiler.cpp and the IR could be a source of bugs.

Reply:
I think this will compile and then fail in schema matching as a result of ^^^

```
def foo():
   a = 1
   for i in range(0):
       a = None
   print(a + a)
```

Reply:
Because store doesn't remember the type of parent variable.
```
a : Optional[Tensor] = torch.ones()
if True:
  a = None
```

The store will store a value of type tensor. It doesn't remember that it is storing optional tensor (afterall, where would it?).  So the convert_to_ssa pass is going to derive type Tensor, but the original script has type Optional[Tensor].

Reply:
Makes sense. another option is making prim::Store remember Optional[Tensor] here